src/cls/rgw/cls_rgw.cc

```diff
@@ -2978,7 +2978,7 @@ static int list_instance_entries(cls_method_context_t hctx,
       return ret;
     }
   }
-  if (found_first) {
+  if (found_first && start_after_key != marker) {
```

Comment:
Thank you for this fix. I think it would be improved if this logic were added to line 2967 when found_first is assigned. That way we won't decrement `max` unnecessarily either. 

```
// we need to include the exact match if a filter (name) is
// specified and the marker has not yet advanced (i.e., been set)
bool found_first = (ret == 0) && (start_after_key != marker);
```

Note 1: the parens are not needed but may add clarity to the logic.
Note 2: the same suggestion applies to `list_olh_entries` also

Reply:
I agree with your suggestion